### PR TITLE
fix: support more combinations of volumes and bind mounts for docker compose transpiler

### PR DESCRIPTION
**Pull request by h4ck3rk3y**
@tedim52  this is super janky at the moment

so the logic is

mount everything to /tmp/volume-name/<folder/file name>

then change the entrypoint + cmd to overwrite the target dir with contents.

appreciate any feedback & improvement to this logic

### CODE REVIEW

File: core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go [hunk 1]
```diff
@@ -466,47 +471,55 @@ func getStarlarkFilesArtifacts(composeVolumes []types.ServiceVolumeConfig, servi
...
+				return nil, map[string]string{}, nil, stacktrace.NewError(
 					"Volume path '%v' uses '%v', likely referencing an upstream path on a filesystem. Currently, Kurtosis does not support uploading from host filesystem. "+
 						"Place the contents of '%v' directory inside the package where the compose yaml exists and update the volume filepath to be a relative path within the package.",
 					volume.Source, upstreamRelativePathSymbol, volume.Source)
 			}
 
 			// Assume that if an absolute path is specified, user wants to use volume as a persistence layer
 			// Additionally, assume relative paths are read-only
+			// GM - this should fail too as this could be outside the repository
 			shouldPersist = path.IsAbs(volume.Source)
```
Comment (tedim52) [inline core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go]: we are making the assumption that they want to use volume as persistence layer, so we create a persistent directory with `volume.Source` as the dir path on the service if they ref an absolute path on host machine - thus, we don't actually read anything from the filesystem. Maybe we can make this assumption/case more explicit

Comment (h4ck3rk3y) [reply core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go]: Have removed my comment!

File: container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go [hunk 2]
```diff
@@ -1660,6 +1660,14 @@ func (manager *DockerManager) getImagePlatform(ctx context.Context, imageName st
 	return imageInspect.Architecture, nil
 }
 
+func (manager *DockerManager) GetOriginalEntryPointAndCommand(ctx context.Context, imageName string) ([]string, []string, error) {
```
Comment (tedim52) [inline container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go:1663]: as far as docker manager is concerned, we're not changing the entrypoint and command. wdyt abt just`GetEntryPointAndCommand` ?

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/service_config.go [hunk 3]
```diff
@@ -504,6 +514,20 @@ func (config *ServiceConfig) ToKurtosisType(
...
+		return nil, interpretationErr
+	}
+	if found && filesToBeMovedStarlark.Len() > 0 {
+		logrus.Info("this is confusing")
+		filesToBeMoved, interpretationErr = kurtosis_types.SafeCastToMapStringString(filesToBeMovedStarlark, FilesToBeMovedAttr)
+		if interpretationErr != nil {
+			return nil, interpretationErr
+		}
+	}
+	logrus.Infof("the attribute is '%v'", filesToBeMoved)
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/service_config.go:529]: is this log stmt needed?

Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/kurtosis_types/service_config/service_config.go:529]: remove

File: container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go (hunk 2 above)
Comment (h4ck3rk3y) [reply container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go:1663]: ok

File: core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/service_config_full_framework_test.go [hunk 4]
```diff
@@ -39,7 +39,7 @@ func (t *serviceConfigFullTestCase) GetStarlarkCode() string {
 	fileArtifact1 := fmt.Sprintf("%s(%s=[%q])", directory.DirectoryTypeName, directory.ArtifactNamesAttr, testFilesArtifactName1)
 	fileArtifact2 := fmt.Sprintf("%s(%s=[%q])", directory.DirectoryTypeName, directory.ArtifactNamesAttr, testFilesArtifactName2)
 	persistentDirectory := fmt.Sprintf("%s(%s=%q)", directory.DirectoryTypeName, directory.PersistentKeyAttr, testPersistentDirectoryKey)
-	starlarkCode := fmt.Sprintf("%s(%s=%q, %s=%s, %s=%s, %s=%s, %s=%s, %s=%s, %s=%s, %s=%q, %s=%d, %s=%d, %s=%d, %s=%d, %s=%s, %s=%v, %s=%v)",
+	starlarkCode := fmt.Sprintf("%s(%s=%q, %s=%s, %s=%s, %s=%s, %s=%s, %s=%s, %s=%s, %s=%q, %s=%d, %s=%d, %s=%d, %s=%d, %s=%s, %s=%v, %s=%v, files_to_be_moved={\"/foo/bar\": \"/doo/dar\"})",
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/service_config_full_framework_test.go:42]: can we follow same pattern as below? `%s=%v` and add as args

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/start_user_services.go [hunk 5]
```diff
@@ -552,6 +556,18 @@ func createStartServiceOperation(
 			envVars[key] = strings.Replace(envVars[key], privateIPAddrPlaceholder, privateIPAddrStr, unlimitedReplacements)
 		}
 
+		// THIS IS A HACK
```
Comment (tedim52) [inline container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/start_user_services.go:559]: i like that the hack is isolated to one function and compose transpilation this is good

**Review by tedim52 — APPROVED** [review]
ive made changes to this logic but those will be reflected in my PR

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/start_user_services.go [hunk 6]
```diff
@@ -753,6 +769,51 @@ func createStartServiceOperation(
 	}
 }
 
+// TODO - clean this up this is super janky, a way to handle compose  & volume
+func handleFilesToBeMovedForDockerCompose(ctx context.Context, dockerManager *docker_manager.DockerManager, containerImageName string, cmdArgs []string, entrypointArgs []string, filesToBeMoved map[string]string) ([]string, []string, error) {
```
Comment (tedim52) [inline container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/start_user_services.go:773]: extremely nit pick but can we keep reference to `DockerCompose` in this module to components, ideally the code shouldn't reflect any knowledge of DockerCompose even though thats the purpose of the hack, especially if this  this hack remains in the codebase for some reason)

Comment (tedim52) [reply container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/start_user_services.go:773]: `s/handleFilesToBeMovedForDockerCompose/handleFilesToBeMoved`

**Comment by tedim52** [comment]
this can be merged even with the git guardian secrets error - i need to make a PR to update our git history to remove the fake env var PASSWORD secrets I used in testing but there are no actual secrets in git history

File: core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/service_config_full_framework_test.go (hunk 4 above)
Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/service_config_full_framework_test.go:42]: sure that works for me

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/start_user_services.go (hunk 6 above)
Comment (h4ck3rk3y) [reply container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/start_user_services.go:773]: that makes sense to me

**Comment by tedim52** [comment]
merged in https://github.com/kurtosis-tech/kurtosis/pull/2320

**Pull request closed without merging**